### PR TITLE
Add maxPool, moments, stopGradient, sin, cos, tan, sqrt, pow, and new layers api with batchNorm, conv2d, linear

### PR DESCRIPTION
**Pull request by ry**

### CODE REVIEW

File: src/ops.ts [hunk 1]
```diff
@@ -276,6 +276,21 @@ defBW("square", (g, x) => {
   return g.mul(x.mul(two));
 });
 
+export const pow = defFW("pow", (x: types.Storage, e: types.Storage) => {
+  return bo.pow(x, e);
+});
+defBW("pow", (g: Tensor) => {
+  throw Error("Not Implemented.");
```
Comment (piscisaureus) [inline src/ops.ts]: g.mul(e).mul(x.pow(e.sub(1))) I think

File: src/dl/math/backends/webgl/binaryop_gpu.ts [hunk 2]
```diff
@@ -28,8 +28,9 @@ export const SUB = "return a - b;";
 export const MUL = "return a * b;";
 export const DIV = "return a / b;";
 export const POW = `
-  return (round(mod(b, 2.0)) == 0 || round(mod(b, 2.0)) == 2) ?
-      pow(abs(a), b) : sign(a) * pow(abs(a), b);
+  return (a >= 0.) ? pow(a, b) :
+    ((round(mod(b, 2.0)) == 0 || round(mod(b, 2.0)) == 2) ?
```
Comment (piscisaureus) [inline src/dl/math/backends/webgl/binaryop_gpu.ts]: I don't think this works properly when `a < 0` and b is not a round number.

File: src/experiment.ts [hunk 3]
```diff
@@ -169,7 +169,9 @@ export abstract class Experiment {
     gc((keep) => {
       const gradFn = gradParams(lossFn);
       // Forward/Backward pass
+      (this.currentParams as any).isTraining = true;
```
Comment (piscisaureus) [inline src/experiment.ts]: This cast shouldn't be necessary.

File: src/experiment.ts [hunk 4]
```diff
@@ -169,7 +169,9 @@ export abstract class Experiment {
...
       const gradsAndLoss = gradFn(this.currentParams);
+      (this.currentParams as any).isTraining = false;
```
Comment (piscisaureus) [inline src/experiment.ts]: ditto

File: src/ops.ts (hunk 1 above)
Comment (ry) [reply src/ops.ts]: thanks. fixed.

File: tools/gendoc.ts [hunk 5]
```diff
@@ -188,6 +188,10 @@ export function genJSON(): DocEntry[] {
       // TODO Ignoring for now.
       log("- ObjectLiteralExpression");
 
+    } else if (ts.isTypeLiteralNode(node)) {
+      // TODO Ignoring for now.
+      log("- TypeLiteral");
```
[the remark this answers is not in the document]
Comment (ry) [reply tools/gendoc.ts]: Not sure exactly.. : )

File: src/tensor.ts [hunk 6]
```diff
@@ -388,6 +388,27 @@ export class Tensor implements types.Storage {
     return ops.reduceLogSumExp(this, axes, keepDims);
   }
 
+  /** Prevents from backproping thru the tensor. Returned tensor uses the same
+   * storage.
+   */
+  stopGradient(): Tensor {
```
[the remark this answers is not in the document]
Comment (ry) [reply src/tensor.ts]: So, those other methods are internal ops - where I think Grad is shorter and consistent with previous gradient ops like reluGrad. This is meant to be familiar to TF users https://www.tensorflow.org/api_docs/python/tf/stop_gradient

File: src/api.ts [hunk 7]
```diff
@@ -183,3 +184,29 @@ export function int32(t: types.TensorLike): Tensor {
 export function float32(t: types.TensorLike): Tensor {
   return tensor(t, { dtype: "float32" });
 }
+
+/** Performs 2d convolution.
+ * The input and filter tensors should both be rank 4 and float32, with the
+ * input formatted as [batch, height, width, channels] and the filter
+ * [height, width, in chans, out chans].
+ *
+ * Additional arguments are ConvOpts { stride, padding, bias }.
```
[the remark this answers is not in the document]
Comment (ry) [reply src/api.ts]: Fixed

File: src/dl/math/backends/webgl/binaryop_gpu.ts (hunk 2 above)
Comment (ry) [reply src/dl/math/backends/webgl/binaryop_gpu.ts]: Yeah, I'm not sure on which real numbers this is defined, at least a = -1 and b = 1/2 is imaginary ... I'd leave it as until we encounter counter-examples

File: src/experiment.ts (hunk 3 above)
Comment (ry) [reply src/experiment.ts]: Fixed.

File: src/experiment.ts (hunk 4 above)
Comment (ry) [reply src/experiment.ts]: Fixed.
